src/coordinator/tsdb/remote/compressed_codecs.go

```diff
@@ -0,0 +1,298 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package remote
+
+import (
+	"io"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3db/src/coordinator/generated/proto/rpc"
+	"github.com/m3db/m3db/src/dbnode/encoding"
+	"github.com/m3db/m3db/src/dbnode/encoding/m3tsz"
+	"github.com/m3db/m3db/src/dbnode/ts"
+	"github.com/m3db/m3db/src/dbnode/x/xio"
+	"github.com/m3db/m3db/src/dbnode/x/xpool"
+	"github.com/m3db/m3x/checked"
+	"github.com/m3db/m3x/ident"
+	xtime "github.com/m3db/m3x/time"
+)
+
+func initializeVars() {
+	opts = checked.NewBytesOptions().SetFinalizer(
+		checked.BytesFinalizerFn(func(b checked.Bytes) {
+			b.Reset(nil)
+		}))
+
+	iterAlloc = func(r io.Reader) encoding.ReaderIterator {
+		return m3tsz.NewReaderIterator(r, m3tsz.DefaultIntOptimizationEnabled, encoding.NewOptions())
+	}
+}
+
+var (
+	opts       checked.BytesOptions
+	iterAlloc  func(r io.Reader) encoding.ReaderIterator
+	initialize sync.Once
+)
+
+func compressedSegmentFromBlockReader(br xio.BlockReader) (*rpc.Segment, error) {
+	segment, err := br.Segment()
+	if err != nil {
+		return nil, err
+	}
+	return &rpc.Segment{
+		Head:      segment.Head.Bytes(),
+		Tail:      segment.Tail.Bytes(),
+		StartTime: xtime.ToNanoseconds(br.Start),
+		BlockSize: int64(br.BlockSize),
+	}, nil
+}
+
+func compressedSegmentsFromReaders(readers xio.ReaderSliceOfSlicesIterator) (*rpc.Segments, error) {
+	segments := &rpc.Segments{}
+	l, _, _ := readers.CurrentReaders()
+	// NB(arnikola) If there's only a single reader, the segment has been merged
+	// otherwise, multiple unmerged segments exist.
+	if l == 1 {
+		br := readers.CurrentReaderAt(0)
+		segment, err := compressedSegmentFromBlockReader(br)
+		if err != nil {
+			return nil, err
+		}
+		segments.Merged = segment
+	} else {
+		unmerged := make([]*rpc.Segment, 0, l)
+		for i := 0; i < l; i++ {
+			br := readers.CurrentReaderAt(i)
+			segment, err := compressedSegmentFromBlockReader(br)
+			if err != nil {
+				return nil, err
+			}
+			unmerged = append(unmerged, segment)
+		}
+		segments.Unmerged = unmerged
+	}
+	return segments, nil
+}
+
+func compressedTagsFromTagIterator(tagIter ident.TagIterator) ([]*rpc.CompressedTag, error) {
+	tags := make([]*rpc.CompressedTag, 0, tagIter.Remaining())
+	for tagIter.Next() {
+		tag := tagIter.Current()
+		tags = append(tags, &rpc.CompressedTag{
+			Name:  tag.Name.Bytes(),
+			Value: tag.Value.Bytes(),
+		})
+	}
+	err := tagIter.Err()
+	if err != nil {
+		return nil, err
+	}
+	return tags, nil
+}
+
+// CompressedProtobufFromSeriesIterator builds compressed rpc series from a SeriesIterator
+func CompressedProtobufFromSeriesIterator(it encoding.SeriesIterator) (*rpc.Series, error) {
```

Comment:
can you add some high level comments on the steps involved to convert SeriesIterator to rpc.Series ?